src/box-client.ts

```diff
@@ -174,182 +173,643 @@ function formatResponseForBatch(response) {
+	 */
+	setIPs(ips: string[]) {
+		var validIPs = ips.filter((ipString: string) => isIP(ipString)).join(', ');
 
-							var expiredTokensError = errors.buildAuthError(response);
+		this.setCustomHeader(HEADER_XFF, validIPs);
 
-							// Give the session a chance to handle the error (ex: a persistent session will clear the token store)
-							if (this._session.handleExpiredTokensError) {
-								this._session.handleExpiredTokensError(expiredTokensError);
-							}
-						}
+		this._tokenOptions = { ip: validIPs };
+	}
+
+	/**
+	 * Sets the shared item context on the API Session. Overwrites any current context.
+	 *
+	 * @param {string} url The shared link url
+	 * @param {?string} password The shared link password, null if no password exists.
+	 * @returns {void}
+	 */
+	setSharedContext(url: string, password: string | null) {
+		var sharedContextAuthHeader = this.buildSharedItemAuthHeader(url, password);
+		this.setCustomHeader(HEADER_BOXAPI, sharedContextAuthHeader);
+	}
+
+	/**
+	 * Removes any current shared item context from API Session.
+	 *
+	 * @returns {void}
+	 */
+	revokeSharedContext() {
+		this.setCustomHeader(HEADER_BOXAPI, null);
+	}
+
+	/**
+	 * Set up the As-User context, which is used by enterprise admins to
+	 * impersonate their managed users and perform actions on their behalf.
+	 *
+	 * @param {string} userID - The ID of the user to impersonate
+	 * @returns {void}
+	 */
+	asUser(userID: string) {
+		this.setCustomHeader(HEADER_AS_USER, userID);
+	}
+
+	/**
+	 * Revoke the As-User context and return to making calls on behalf of the user
+	 * who owns the client's access token.
+	 *
+	 * @returns {void}
+	 */
+	asSelf() {
+		this.setCustomHeader(HEADER_AS_USER, null);
+	}
+
+	/**
+	 * Revokes the client's access tokens. The client will no longer be tied to a user
+	 * and will be unable to make calls to the API, rendering it effectively useless.
+	 *
+	 * @param {Function} [callback] Called after revoking, with an error if one existed
+	 * @returns {Promise} A promise resolving when the client's access token is revoked
+	 */
+	revokeTokens(callback: Function) {
+		return this._session.revokeTokens(this._tokenOptions).asCallback(callback);
+	}
+
+	/**
+	 * Exchange the client access token for one with lower scope
+	 * @param {string|string[]} scopes The scope(s) requested for the new token
+	 * @param {string} [resource] The absolute URL of an API resource to scope the new token to
+	 * @param {Object} [options] - Optional parameters
+	 * @param {ActorParams} [options.actor] - Optional actor parameters for creating annotator tokens with Token Auth client
+	 * @param {SharedLinkParams} [options.sharedLink] - Optional shared link parameters for creating tokens using shared links
+	 * @param {Function} [callback] Called with the new token
+	 * @returns {Promise<TokenInfo>} A promise resolving to the exchanged token info
+	 */
+	exchangeToken(
+		scopes: string | string[],
+		resource: string,
+		options: Object,
```

Comment:
Looks like `options` can sometimes be a function as well, looking at line 508 below